prow/plugins/plugins.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugins
+
+import (
+	"fmt"
+	"io/ioutil"
+	"sync"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/ghodss/yaml"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/jobs"
+	"k8s.io/test-infra/prow/kube"
+)
+
+var (
+	allPlugins           = map[string]struct{}{}
+	issueCommentHandlers = map[string]IssueCommentHandler{}
+	pullRequestHandlers  = map[string]PullRequestHandler{}
+)
+
+type IssueCommentHandler func(*PluginAgent, github.IssueCommentEvent) error
+
+func RegisterIssueCommentHandler(name string, fn IssueCommentHandler) {
+	allPlugins[name] = struct{}{}
+	issueCommentHandlers[name] = fn
+}
+
+type PullRequestHandler func(*PluginAgent, github.PullRequestEvent) error
+
+func RegisterPullRequestHandler(name string, fn PullRequestHandler) {
+	allPlugins[name] = struct{}{}
+	pullRequestHandlers[name] = fn
+}
+
+type PluginAgent struct {
```

Comment:
I wish you didn't have to pass this entire type to each plugin. `GitHubClient` seems useful for most, it's much less true of `KubeClient` or `JobAgent`, not super critical.